singularity-opac/neutrinos/neutrino_s_variant.hpp

```diff
@@ -0,0 +1,141 @@
+#include <singularity-opac/base/radiation_types.hpp>
+#include <variant/include/mpark/variant.hpp>
+
+namespace singularity {
+namespace neutrinos {
+namespace impl {
+
+template <typename... Ts>
+using s_opac_variant = mpark::variant<Ts...>;
+
+template <typename... S_Opacs>
+class S_Variant {
+ private:
+  s_opac_variant<S_Opacs...> s_opac_;
+
+ public:
+  template <
+      typename Choice,
+      typename std::enable_if<
+          !std::is_same<S_Variant, typename std::decay<Choice>::type>::value,
+          bool>::type = true>
+  PORTABLE_FUNCTION S_Variant(Choice &&choice)
+      : s_opac_(std::forward<Choice>(choice)) {}
+
+  PORTABLE_FUNCTION
+  S_Variant() noexcept = default;
+
+  template <
+      typename Choice,
+      typename std::enable_if<
+          !std::is_same<S_Variant, typename std::decay<Choice>::type>::value,
+          bool>::type = true>
+  PORTABLE_FUNCTION S_Variant &operator=(Choice &&s_opac) {
+    s_opac_ = std::forward<Choice>(s_opac);
+    return *this;
+  }
+
+  template <
+      typename Choice,
+      typename std::enable_if<
+          !std::is_same<S_Variant, typename std::decay<Choice>::type>::value,
+          bool>::type = true>
+  Choice get() {
+    return mpark::get<Choice>(s_opac_);
+  }
+
+  S_Variant GetOnDevice() {
+    return mpark::visit(
+        [](auto &s_opac) {
+          return s_opac_variant<S_Opacs...>(s_opac.GetOnDevice());
+        },
+        s_opac_);
+  }
+
+  // Total cross section with units of length^2
+  // Signature should be at least
+  // rho, temp, Ye, type, nu, lambda
+  PORTABLE_INLINE_FUNCTION Real TotalCrossSection(
+    const Real rho, const Real temp, const Real Ye, const RadiationType type,
+    const Real nu, Real *lambda = nullptr) const {
+      return mpark::visit(
+       [=](const auto &s_opac) {
+         return s_opac.TotalCrossSection(rho, temp, Ye, type, nu, lambda);
+       }, s_opac_);
+    }
+
+  // Differential cross section with units of length^2/steradian
+  // Signature should be at least
+  // rho, temp, Ye, type, nu, theta, lambda
+  PORTABLE_INLINE_FUNCTION Real DifferentialCrossSection(
+    const Real rho, const Real temp, const Real Ye, const RadiationType type,
+    const Real nu, const Real theta, Real *lambda = nullptr) const {
+      return mpark::visit(
+      [=](const auto &s_opac) {
+        return s_opac.DifferentialCrossSection(rho, temp, Ye, type, nu, theta, lambda);
+      }, s_opac_);
+    }
+
+  // Total scattering coefficient with units of 1/length
+  // Signature should be at least
+  // rho, temp, Ye, type, nu, lambda
+  PORTABLE_INLINE_FUNCTION Real TotalScatteringCoefficient(
+      const Real rho, const Real temp, const Real Ye, const RadiationType type,
+      const Real nu, Real *lambda = nullptr) const {
+    return mpark::visit(
+        [=](const auto &s_opac) {
+          return s_opac.ScatteringCoefficient(rho, temp, Ye, type, nu, lambda);
+        },
+        s_opac_);
+  }
```

Comment:
@Yurlungur How about these new functions:
```c++
Real TotalCrossSection(
    const Real rho, const Real temp, const Real Ye, const RadiationType type,
    const Real nu, Real *lambda = nullptr);
Real DifferentialCrossSection(
    const Real rho, const Real temp, const Real Ye, const RadiationType type,
    const Real nu, const Real theta, Real *lambda = nullptr);
Real TotalScatteringCoefficient(
      const Real rho, const Real temp, const Real Ye, const RadiationType type,
      const Real nu, Real *lambda = nullptr);
```

Reply:
that works. Is there a reason to use theta vs mu? Probably doesn't matter much, but mu might let us offload some of the work of computing transcendental functions

Reply:
Yeah I think `theta` will always appear inside a trigonometric function so if the code already has `mu` there is no reason to have them turn it into ` theta` before we turn it back into a `mu` or `1 - mu^2`.

Reply:
EDIT: I did indeed make this change `theta` -> `mu`